### ISSUE TITLE
Proposal: Replace anytype with specially marked generic variables

### ISSUE THREAD

**Issue by zzyxyzz** [ziglang/zig#9244]
## Proposal

In place of `anytype`, introduce the concept of generic/dynamic variables, which are declared by prefixing the name with `$` and not giving a type signature. This syntax would be allowed in function declarations and comptime fields:

```zig
fn eql($a, b: @TypeOf(a)) bool {
    return a == b;
}

const foo = struct {
    $bar = .{},
};

// Optional extension (currently not supported by anytype):
comptime var $x = 1;
x = "hello";
```

## Rationale / Advantages over anytype

The problem that this proposal addresses is merely aesthetic, but one that the community has wrangled with repeatedly: Originally, generic fields and function parameters were called `var`, which was actually reasonably fitting. However, because of the name clash with variable declaration and possible allusion to varargs, it was renamed to `anytype` (#4820). This, however, is deemed by many to be at least as confusing (#5893), but even after extensive bikeshedding no alternative name seems to have reached critical concensus mass. In addition, `anytype`, despite occuring in the type signature, is not really a type (#9107), but a keyword with special meaning. (E.g., you cannot construct a `?anytype`, etc. It is also the only "type" for which `x: T` -> `@TypeOf(x)==T` is *not* true.)

Instead of using a special keyword in lieu of a type, the intent to infer the type from the supplied value may be better signaled by omitting the type signature entirely. This is how most languages with optional type annotations handle it. In Zig, the same thing happens when you write `const x = "hello";` This usage could be extended to include function arguments and struct fields as well:

```zig
fn eql(a, b: @TypeOf(a)) bool {
    return a == b;
}

const foo = struct {
    bar,
};
```

This solution was actually suggested early in the discussion surrounding `var` and `anytype` by @andrewrk ([4820](https://github.com/ziglang/zig/issues/4820#issuecomment-604607623)) with the following comment:

> The entire : foo syntax is redundant, we could even just have it so that you could omit the type and it would be any type. The counter-argument there was that it would encourage people to make generic functions even where generic functions are not appropriate.

Apart from encouraging the overuse of generic functions, we would also lose the ability to distinguish forgotten and deliberately omitted type signatures, which seems like a major loss for error reporting. Thus the need to add some other mark to the declaration. `fn (generic a, ...` was recently suggested by @YohananDiamond ([5893 (comment)](https://github.com/ziglang/zig/issues/5893#issuecomment-854243446)). But for myself I prefer a sigil as proposed above, for the simple reason that it doesn't carry any linguistic baggage and we don't have to restart the discussion of what `generic` or `any` or `inferred` means to whom :smile:.

## Open questions

I'm not sure what the type signature of a generic function would be. `fn(anytype)`? `fn(_)`? Other suggestions?

**Comment by batiati** [comment]
IMO `anytype` is fine, I just would like if something like that were possible:

```zig
fn foo(a: trait.IsTuple(@TypeOf(a))) void {
   // here I know that @TypeOf(a) is a tuple and was validated on the signature
}

fn my_iter(iter: someCustomCheck(@TypeOf(iter)) ElementOf(@TypeOf(iter)) {
  // someCustomCheck is a comptime function that evaluates the type
  // ElementOf(@TypeOf(iter)) is possible on status-quo zig
}
```

Putting all traits validations on the signature can improve the API clarity over status-quo comptime checks inside the method's body and `anytype` + comments on the signature.

**Comment by zzyxyzz** [comment]
@batiati
That would indeed be useful, but it's mostly a different issue (#6615).

**Comment by YohananDiamond** [comment]
> Apart from encouraging the overuse of generic functions, we would also lose the ability to distinguish forgotten and deliberately omitted type signatures, which seems like a major loss for error reporting. Thus the need to add some other mark to the declaration. `fn (generic a, ...` was recently suggested by @YohananDiamond (5893 (comment)). [...]

Hm... in my example, the syntax I proposed was actually `fn (generic a: _, ...` - the previous one was part of my line of thinking - so I'm not sure if you meant the one with the `_` or not.
Would a `fn (parameter_name: _, ...` syntax solve this or do you think it ends up at the same problem?

By the way, there's something I ended up changing from my original idea - since `anytype` on structs and function parameters mean different things (compile-time dynamic variables and function parameters, respectively), maybe they should have different syntaxes.

**Comment by zzyxyzz** [comment]
@YohananDiamond
Sorry if I misrepresented the intent of your proposed solution. I was specifically referring to the `fn (generic a, generic b, ...)` syntax, but maybe that was not your preferred option.

> Would a fn (parameter_name: _, ... syntax solve this or do you think it ends up at the same problem?

The `x: _` syntax ([#5893 (comment)](https://github.com/ziglang/zig/issues/5893#issuecomment-769027502)), is actually one of my favorites. In the proposal here I tried to make the case that the type signature should be omitted entirely, if we don't intend to put type constraints on a parameter or field. But if that doesn't gain traction, then my vote would be for `x: _`.
